architecture.adoc

```diff
@@ -0,0 +1,134 @@
+= Architecture
+
+= Introduction
+
+Mokey is an application that collects data about running applications on an Aurora OpenShift cluster and exposes that
+information via a ReST API in a cached and fast-to-access manner.
+
+The collection process is scheduled and parallelized, and information about the applications is updated in real time as
+the process progresses.
+
+This document is a description of the overall architecture of the module, divided into two main sections;
+<<core-concepts, the Core Concepts>> and <<api-resources, the API resources>>, describing the internal aspects of the
+module, and the externally exposed resources, respectively.
+
+
+[[core-concepts]]
+= Core Concepts (Internal)
+
+This section contains a description of the core concepts of the module, with implementation and domain model details.
+It does not contain any details of the resources that are exposed from the API, only the classes and objects that are
+used to support the internal data collection workflow; crawling.
+
+[[reference-data]]
+== Reference Data
+
+Classes and objects that represent static data and rules.
+
+Package: no.skatteetaten.aurora.mokey.model
+
+[[non-functional-requirement]]NonFunctionalRequirement (NFR):: A description of a non functional requirement for applications running on Aurora
+OpenShift.
+
+NfrViolationCause:: Maybe declare causes for NFR violations?
+
+NfrViolationFix:: Maybe declare fixes for NFR violations?
+
+
+[[application-data]]
+== Application Data
+
```

Comment:
In general i do not like the term data. Because it means everything. I agree that we should use a secondary term after Application but I think data is to broad. 

State
Status

Are these better?

Reply:
They are better if they are accurate ;)

Reply:
Let me try to provide a different view on this. I think ApplicationData (or whatever the name is) is really a DTO that is built to be exposed with a rest service. It is specifically created for a scenario where it all of this information is required.

First I agree with @bjartek that the name isnt the most precise, but its difficult to find a better one. Second, I think we should be very clear that this is an external resource. Perhaps internally this is two entites? For me, when I struggle to find a good name I have usually not found the correct abstraction. If it is two entities, this can be merged into one view model before exposed to the clients.

Reply:
@jarlehansen actually, the opposite was my intent; ApplicationData is primarily in internal representation of data collected for an application intended to be used for several purposes; determining AuroraStatus, determining NfrViolations and supporting the API with exposing potentially several resources. The only reason we need ApplicationData is to increase performance. If reading this data directly from the primary source was fast enough, this class would not be needed.

In my opinion, from ApplicationData we can calculate AuroraStatus, and potentially expose that as an individual resource. We can calculate NfrViolations, and potentially expose those as details in an NfRAuditReport resource (for instance). We can also expsose Application (with few details; name, environment, affiliation...), and potentially an ApplicationDeploymentDetails resource with information about provisioned resources, OpenShift details like pods etc.

Reply:
Ok, understood.

I am just trying to think what is the root entity in this case. It is really an Application, right? and there will be other nested entities tied to this application.

Reply:
Yes, that is true. The Application resource is the "top level" resource and owner of AuroraStatus and NfrAuditReport(? to be determined). I think it would be my preference, though, to model AuroraStatus and NfrAuditReport as top level resources as well, where Application knows how to reference (with links?) the latest instance of these resources (would we want an NfrAuditReport history, for instance?). This way we really do not have to make the distinction about what information is public (for the wallboard) and what requires authentication (for AuroraConsole). I think it would be good if we could the design the resources so that we did not have a PublicApplication and an "Internal"Application resource both exposed in the API.